### ISSUE TITLE
Linking failed when using conan as package manager

### ISSUE THREAD

**Issue by karan-k-deepr** [aws/aws-sdk-cpp#1724]
Confirm by changing [ ] to [x] below to ensure that it's a bug:
- [x] I've gone though [Developer Guide](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/welcome.html) and [API reference](http://sdk.amazonaws.com/cpp/api/LATEST/index.html)
- [x] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-cpp/issues) and didn't find any solution
  
**Describe the bug**
I am using Conan as a package manager in my project on Windows as well as Linux. But for some reason code is not compiling on Windows but the same code is compiling and working perfectly on LINUX platform. 
Previously I tried exactly the same code with manually compiled `aws-sdk-cpp (version: 1.9.49)` works perfectly on windows. 

For installing Conan dependencies in my project I am using `consn install . --profile=conan_profile_for_windows.txt --build=missing -u` . And this command works perfectly without showing any error. 

**SDK version number**
Version: 1.8.130

**Platform/OS/Hardware/Device**
Windows 10 (x86_64)

**To Reproduce (observed behavior)**
using these two `conanfile` and `conanprofile` to install dependencies within my project.
conan_profile_for_windows.txt
```
[settings]
arch=x86_64
arch_build=x86_64
build_type=Debug
compiler=Visual Studio
compiler.runtime=MDd
compiler.version=15
compiler.cppstd=17
compiler.toolset=v141
os=Windows
os_build=Windows
[options]
[build_requires]
[env]
```
conanfile.txt
```
[requires]
aws-sdk-cpp/1.8.130

[generators]
cmake

[options]
aws-sdk-cpp:shared=False
aws-sdk-cpp:cognito-idp=True
aws-sdk-cpp:cognito-sync=True
```

**Expected behavior**

**Logs/output**
```
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl Aws::Client::AWSError<enum Aws::CognitoIdentityProvider::CognitoIdentityProviderErrors>::AWSError<enum Aws::CognitoIdentityProvider::CognitoIdentityProviderErrors>(class Aws::Client::AWSError<enum Aws::CognitoIdentityProvider::CognitoIdentityProviderErrors> const &)" (__imp_??0?$AWSError@W4CognitoIdentityProviderErrors@CognitoIdentityProvider@Aws@@@Client@Aws@@QEAA@AEBV012@@Z) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::isUserValid(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)const " (?isUserValid@CognitoAuth@auth@TestProject@@UEBA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const & __cdecl Aws::Client::AWSError<enum Aws::CognitoIdentityProvider::CognitoIdentityProviderErrors>::GetMessage(void)const " (__imp_?GetMessage@?$AWSError@W4CognitoIdentityProviderErrors@CognitoIdentityProvider@Aws@@@Client@Aws@@QEBAAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::isUserValid(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)const " (?isUserValid@CognitoAuth@auth@TestProject@@UEBA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const & __cdecl Aws::CognitoIdentityProvider::Model::AttributeType::GetName(void)const " (__imp_?GetName@AttributeType@Model@CognitoIdentityProvider@Aws@@QEBAAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::isUserValid(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)const " (?isUserValid@CognitoAuth@auth@TestProject@@UEBA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const & __cdecl Aws::CognitoIdentityProvider::Model::AttributeType::GetValue(void)const " (__imp_?GetValue@AttributeType@Model@CognitoIdentityProvider@Aws@@QEBAAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::isUserValid(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)const " (?isUserValid@CognitoAuth@auth@TestProject@@UEBA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const & __cdecl Aws::CognitoIdentityProvider::Model::AuthenticationResultType::GetAccessToken(void)const " (__imp_?GetAccessToken@AuthenticationResultType@Model@CognitoIdentityProvider@Aws@@QEBAAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::login(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?login@CognitoAuth@auth@TestProject@@UEAA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@0@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: int __cdecl Aws::CognitoIdentityProvider::Model::AuthenticationResultType::GetExpiresIn(void)const " (__imp_?GetExpiresIn@AuthenticationResultType@Model@CognitoIdentityProvider@Aws@@QEBAHXZ) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::login(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?login@CognitoAuth@auth@TestProject@@UEAA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@0@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const & __cdecl Aws::CognitoIdentityProvider::Model::AuthenticationResultType::GetRefreshToken(void)const " (__imp_?GetRefreshToken@AuthenticationResultType@Model@CognitoIdentityProvider@Aws@@QEBAAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::login(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?login@CognitoAuth@auth@TestProject@@UEAA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@0@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const & __cdecl Aws::CognitoIdentityProvider::Model::AuthenticationResultType::GetIdToken(void)const " (__imp_?GetIdToken@AuthenticationResultType@Model@CognitoIdentityProvider@Aws@@QEBAAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::login(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?login@CognitoAuth@auth@TestProject@@UEAA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@0@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl Aws::CognitoIdentityProvider::Model::AuthenticationResultType::AuthenticationResultType(class Aws::CognitoIdentityProvider::Model::AuthenticationResultType const &)" (__imp_??0AuthenticationResultType@Model@CognitoIdentityProvider@Aws@@QEAA@AEBV0123@@Z) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::login(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?login@CognitoAuth@auth@TestProject@@UEAA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@0@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const & __cdecl Aws::CognitoIdentityProvider::Model::GetUserResult::GetUsername(void)const " (__imp_?GetUsername@GetUserResult@Model@CognitoIdentityProvider@Aws@@QEBAAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::isUserValid(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)const " (?isUserValid@CognitoAuth@auth@TestProject@@UEBA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class std::vector<class Aws::CognitoIdentityProvider::Model::AttributeType,class std::allocator<class Aws::CognitoIdentityProvider::Model::AttributeType> > const & __cdecl Aws::CognitoIdentityProvider::Model::GetUserResult::GetUserAttributes(void)const " (__imp_?GetUserAttributes@GetUserResult@Model@CognitoIdentityProvider@Aws@@QEBAAEBV?$vector@VAttributeType@Model@CognitoIdentityProvider@Aws@@V?$allocator@VAttributeType@Model@CognitoIdentityProvider@Aws@@@std@@@std@@XZ) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::isUserValid(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)const " (?isUserValid@CognitoAuth@auth@TestProject@@UEBA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: enum Aws::CognitoIdentityProvider::Model::ChallengeNameType const & __cdecl Aws::CognitoIdentityProvider::Model::InitiateAuthResult::GetChallengeName(void)const " (__imp_?GetChallengeName@InitiateAuthResult@Model@CognitoIdentityProvider@Aws@@QEBAAEBW4ChallengeNameType@234@XZ) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::login(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?login@CognitoAuth@auth@TestProject@@UEAA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@0@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class Aws::CognitoIdentityProvider::Model::AuthenticationResultType const & __cdecl Aws::CognitoIdentityProvider::Model::InitiateAuthResult::GetAuthenticationResult(void)const " (__imp_?GetAuthenticationResult@InitiateAuthResult@Model@CognitoIdentityProvider@Aws@@QEBAAEBVAuthenticationResultType@234@XZ) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::login(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?login@CognitoAuth@auth@TestProject@@UEAA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@0@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl Aws::CognitoIdentityProvider::Model::InitiateAuthResult::InitiateAuthResult(class Aws::CognitoIdentityProvider::Model::InitiateAuthResult const &)" (__imp_??0InitiateAuthResult@Model@CognitoIdentityProvider@Aws@@QEAA@AEBV0123@@Z) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::login(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?login@CognitoAuth@auth@TestProject@@UEAA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@0@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: void __cdecl Aws::CognitoIdentityProvider::Model::InitiateAuthRequest::SetAuthFlow(enum Aws::CognitoIdentityProvider::Model::AuthFlowType &&)" (__imp_?SetAuthFlow@InitiateAuthRequest@Model@CognitoIdentityProvider@Aws@@QEAAX$$QEAW4AuthFlowType@234@@Z) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::login(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?login@CognitoAuth@auth@TestProject@@UEAA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@0@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: void __cdecl Aws::CognitoIdentityProvider::Model::InitiateAuthRequest::SetAuthParameters(class std::map<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,struct std::less<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >,class std::allocator<struct std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const ,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > > > > const &)" (__imp_?SetAuthParameters@InitiateAuthRequest@Model@CognitoIdentityProvider@Aws@@QEAAXAEBV?$map@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@V12@U?$less@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@2@V?$allocator@U?$pair@$$CBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@V12@@std@@@2@@std@@@Z) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::login(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?login@CognitoAuth@auth@TestProject@@UEAA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@0@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: void __cdecl Aws::CognitoIdentityProvider::Model::InitiateAuthRequest::SetClientId(char const *)" (__imp_?SetClientId@InitiateAuthRequest@Model@CognitoIdentityProvider@Aws@@QEAAXPEBD@Z) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::login(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?login@CognitoAuth@auth@TestProject@@UEAA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@0@Z)
[build] cognito_auth.cpp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: void __cdecl Aws::CognitoIdentityProvider::Model::GetUserRequest::SetAccessToken(char const *)" (__imp_?SetAccessToken@GetUserRequest@Model@CognitoIdentityProvider@Aws@@QEAAXPEBD@Z) referenced in function "public: virtual bool __cdecl TestProject::auth::CognitoAuth::isUserValid(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)const " (?isUserValid@CognitoAuth@auth@TestProject@@UEBA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)

```

**Comment by KaibaLopez** [comment]
Hi @karan-k-deepr ,
So the first thing here is we don't really support connan,  meaning that if there is something that is specifically wrong with using connan we might not fix it.
That being said, I'll give it a shot see if I can figure something out, keep in mind that I'm not an expert using connan so I'll have some, maybe obvious, questions.

> Previously I tried exactly the same code with manually compiled aws-sdk-cpp (version: 1.9.49) works perfectly on windows.

This kind of points at the problem not being the SDK but the connan build ( or linking?) of it.
Also, you say that you manually installed 1.9 but are then trying to use 1.8.130 on connan, any reason for this? 

finally, before I go digging into this, kind of a shot in the dark, but those errors are kind of common when building manually and cmake and MSBuild are out of sync on the build type ( I think Cmake build type defaults to release and MSBuild to Debug ), could you verify that's not the case for you?

**Comment by karan-k-deepr** [comment]
Hello @KaibaLopez ,
Thanks for your reply. Actually, I was smashing my head on my test code why exactly it was working with the manual compiled library but not with what is provided by Conan. 🤔 So I asked this question here maybe it has some relation with the SDK or version which is there in Conan. 

> This kind of points at the problem not being the SDK but the connan build ( or linking?) of it.
Also, you say that you manually installed 1.9 but are then trying to use 1.8.130 on connan, any reason for this?

Previously I was using test_project with manually compiled AWS SDK (`using the latest one 1.9.49`). And then I moved to Conan, which reduces the manual compilation headache. And unfortunately, Conan provides the older version of AWS (`which is 1.8.130`). It would be better if Conan will have the latest version as soon as there is a release on `AWS SDK's` GitHub page. 

Before giving you some reply I tried creating a new test project with Conan and it works fine. The issue I got is with `USE_IMPORT_EXPORT` macro because of some issues I was facing when building project with manually compiled libraries. 
https://stackoverflow.com/a/41703521/2211345 This is from where I copied the solution. But with Conan, this MACRO is causing issues and works perfectly fine when commented out. 😄
